### ISSUE TITLE
[Bug]: User grants are broken

### ISSUE THREAD

**Issue by meyfa** [zitadel/zitadel#7302]
### Preflight Checklist

- [X] I could not find a solution in the documentation, the existing issues or discussions
- [X] I have joined the [ZITADEL chat](https://zitadel.com/chat)

### Environment

Self-hosted

### Version

2.44.0

### Database

PostgreSQL

### Database Version

16.1

### Describe the problem caused by this bug

During one of the recent updates (likely 2.43.5 to 2.44.0), all our user grants suddenly disappeared. However, they seem to still be present in the database somehow. Specifically:

- When logging into OIDC applications, the application doesn't receive any grant information for the user. If "Assert Roles on Authentication" is checked, the user is unable to log in.
- In the administrator UI, when looking at project grants or at user authorizations, the list is empty. (The user definitely had roles assigned there before, and nobody removed them.)
- When trying to add the same role again, an error notification appears: "User grant already exists (V3-XXXXX)" (where XXXXXX represents some ID string).

We are deploying Zitadel via Helm. We have run the setup job again in case any migrations were missed, but it completes successfully and logs don't indicate a problem.

However, Zitadel's logs contain lots of messages of the following form:

```
level=info msg="process events failed" caller="/home/runner/work/zitadel/zitadel/internal/eventstore/handler/v2/handler.go:416" error="sql: transaction has already been committed or rolled back" projection=projections.user_grants4
```

The message is always the same. It appears a few dozen times on startup and then in regular intervals during runtime (every few seconds).

### To reproduce

1. Start with a Zitadel instance on 2.43.5 or below.
2. Assign a project role to a user.
3. Upgrade to 2.44.0.
4. Look at the user's project roles.

### Expected behavior

Users should keep roles between updates. FWIW, I'm pretty sure they still exist since it's impossible to create them again, however, the query seems to fail - or perhaps a migration fails, as indicated by the log message?

### Operating System

Kubernetes 1.28, deployment via Helm Chart

**Comment by mpjijo** [comment]
Yes, I have the same problem after the upgrade process. My `Authorizations` are now empty.

**Comment by landmass-deftly-reptile-budget** [comment]
Hi, can confirm exactly this behaviour. After upgrade to Zitadel version 2.44.0 with Postgres 16.1 some Authorizations of a user are gone within Zitadel. Trying either manually or via terraform to add the user again to the role results in error message that the user is already assigned to that role/authorization. To me this looks like some "disconnect" or "mismatch" between Zitadel and database.

**Comment by livio-a** [comment]
Hey all

Looks like issues of zitadel not fully recomputing the state after an upgrade. With the release of V2.44 we introduced a new flag / config to be able to ensure this in the setup phase:
https://zitadel.com/docs/support/advisory/a10008

Can you try rerunning the setup with the flag enabled?

BTW: more on those phases you can find here: https://zitadel.com/docs/self-hosting/manage/updating_scaling

**Comment by landmass-deftly-reptile-budget** [comment]
@livio-a 
Hi,
I did already try this but nevertheless did it again right now. 
No change happened. Still the same issue.

Started the container with docker-compose and command:
 command: 'setup --masterkey "blablabla" --init-projections --tlsMode external'

**Comment by livio-a** [comment]
Could correlate to an issue we found while investigating https://discord.com/channels/927474939156643850/1200525276858302586

Can you please execute the following on your database. This might give us some insights:
```
begin; set transaction read only;
select * from projections.current_states where projection_name like 'projections.user_grants%';
select count(*) from projections.user_grants3;
select count(*) from projections.user_grants4;
commit;
```

**Comment by landmass-deftly-reptile-budget** [comment]
@livio-a 
Need to say I am not really a database guy so I hope I did this right:

/ $ psql -U zitadel
psql (16.1)
Type "help" for help.

zitadel=# begin; set transaction read only;
BEGIN
SET
zitadel=*# select * from projections.current_states where projection_name like 'projections.user_grants%';
     projection_name      |    instance_id     |         last_updated          |    aggregate_id    | aggregate_type | sequence |          event_date           |     position      | filter_offset 
--------------------------+--------------------+-------------------------------+--------------------+----------------+----------+-------------------------------+-------------------+---------------
 projections.user_grants4 | 249390599122190339 | 2024-01-30 17:01:25.128225+01 | 249391241605677059 | project        |       11 | 2024-01-28 10:01:36.64252+01  |  1706432496.64255 |             1
 projections.user_grants3 | 249390599122190339 | 2024-01-26 09:04:27.818536+01 | 250887613647421443 | usergrant      |        3 | 2024-01-24 13:26:02.927703+01 | 1706099162.928004 |             1
(2 rows)

zitadel=*# select count(*) from projections.user_grants3;
 count 
-------
     7
(1 row)

zitadel=*# select count(*) from projections.user_grants4;
 count 
-------
     5
(1 row)

zitadel=*# commit;
COMMIT
zitadel=# begin; set transaction read only;
BEGIN
SET

**Comment by meyfa** [comment]
> Can you please execute the following on your database

Here you go:

```
zitadel=*# select * from projections.current_states where projection_name like 'projections.user_grants%';
     projection_name      | instance_id |         last_updated          | aggregate_id | aggregate_type | sequence |          event_date           |     position      | filter_offset
--------------------------+-------------+-------------------------------+--------------+----------------+----------+-------------------------------+-------------------+---------------
 projections.user_grants3 | <redacted>  | 2024-01-26 08:51:51.856436+00 |              |                |        0 | 2023-10-19 15:00:39.583377+00 | 1697727639.583377 |             0
(1 row)

zitadel=*# select count(*) from projections.user_grants3;
 count
-------
    15
(1 row)

zitadel=*# select count(*) from projections.user_grants4;
 count
-------
     0
(1 row)
```

Ours looks different than the previous post, with `projections.user_grants4` missing entirely.

**Comment by hifabienne** [comment]
> It would be cool if this bug could be given priority. My authentication infrastructure is currently down due to this issue, and it’s causing some significant disruptions. 😊

We are already working with high priority on this.

**Comment by livio-a** [comment]
We published a new release: https://github.com/zitadel/zitadel/releases/tag/v2.44.1

This includes a fix for projecting those tables (incl. user_grants).
Please make sure to run zitadel with the init-projections flag.

Regarding that flag:
I just noticed, that the docs have been missleading / incomplete.  (fix is comming: https://github.com/zitadel/zitadel/pull/7325)
the flag must be enabled by `--init-projections=true`

**Comment by meyfa** [comment]
Thank you to everyone who worked on this! I am happy to report that the issue is fixed after running the setup command with `--init-projections=true`.

This flag should probably be added to the Helm chart to avoid the same error in the future, and require less manual work. I have opened https://github.com/zitadel/zitadel-charts/issues/168 to track this.

Closing since this is solved, but feel free to re-open if you want.

**Comment by mpjijo** [comment]
Yes, it's functioning well for me. I appreciate the excellent support and dedication of the team especially @livio-a and @hifabienne   . A special thanks to @meyfa for bringing this issue to their attention for us.
